### PR TITLE
parser for DOS extender + LE executables

### PR DESCRIPTION
**Pull request by user7**

### CODE REVIEW

File: src/nc/input/le/LeParser.cpp [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        rewind:
+            in->seek(mz + 2);
+            continue;
+        }
+        char le_sig[2];
+        if (!read(in, le_sig))
+            goto rewind;
+        if (memcmp(le_sig, "LE", 2))
+            goto rewind;
+        header_pos hpos = { .mz = mz, .le = mz + le_off };
```
Comment (yegord) [inline src/nc/input/le/LeParser.cpp]: This is not supported by MSVC 2010.

File: src/nc/input/le/LeParser.cpp [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        if (!in->getChar(&b))
+            break;
+        if (b != 'Z') {
+            a = b;
+            goto again;
+        }
+        qint64 mz = in->pos() - 2;
+        if (!in->seek(mz + 0x3c))
+            break;
+        std::uint32_t le_off;
```
Comment (yegord) [inline src/nc/input/le/LeParser.cpp]: MSVC 2010 does not support fixed-width types in std:: namespace.
Sorry, I misguided you when asked to add std:: prefix (starting to forget MSVC 2010 issues).

File: src/nc/input/le/LeParser.cpp [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    FIELD(std::uint8_t, flags) \
+    FIELD(std::int16_t, srcoff) \
+
+#define EMIT_ALL \
+    EMIT(LE_HEADER, le_header) \
+    EMIT(OBJ_HEADER, obj_header) \
+    EMIT(FIXUP_HEADER, fixup_header) \
+
+// emitting structs
+#define EMIT(fields, name) struct name { fields } __attribute__((packed));
```
Comment (yegord) [inline src/nc/input/le/LeParser.cpp]: __attribute__ is a GCC extension and not supported by MSVC.

File: src/nc/input/le/LeParser.cpp [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+
+template <typename T>
+void fix_byte_order(T &v) {
+    bo.convertFrom(v);
+}
+
+// -----------------------------------------------------------------------------
+
+#define EMIT(fields, name) \
+QString toQString(const name &h) __attribute__((unused)); \
```
Comment (yegord) [inline src/nc/input/le/LeParser.cpp]: __attribute__ is not supported by MSVC.

File: src/nc/input/le/LeParser.cpp [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+            FIXUP_SRC_16BIT_SELECTOR = 2,
+            FIXUP_SRC_16BIT = 5,
+            FIXUP_SRC_32BIT = 7,
+            FIXUP_DST_FLAG_32BIT_TARGET_OFFSET = 0x10,
+            FIXUP_DST_FLAG_16BIT_OBJECT_ID = 0x40,
+        };
+        while (in->pos() < end) {
+            qint64 start_mark = in->pos();
+            auto throw_truncated = [npage, start_mark]() {
+                throw ParseError(tr("Truncated fixup for page %1 at 0x%2").arg(npage).arg(start_mark, 1, 16));
```
Comment (yegord) [inline src/nc/input/le/LeParser.cpp]: MSVC says `tr` belongs to `this`, and you do not capture it.

**Comment by yegord** [comment]
I have made you code build under all supported compilers, see https://github.com/yegord/snowman/tree/le.

What's left is this:
```
[7/17] Building CXX object nc/CMakeFiles/nc.dir/input/le/LeParser.cpp.o
/home/yegor/devel/snowman/src/nc/input/le/LeParser.cpp:161:1: warning: unused function 'toQString' [-Wunused-function]
EMIT_ALL
^
/home/yegor/devel/snowman/src/nc/input/le/LeParser.cpp:125:5: note: expanded from macro 'EMIT_ALL'
    EMIT(OBJ_HEADER, obj_header) \
    ^
/home/yegor/devel/snowman/src/nc/input/le/LeParser.cpp:155:9: note: expanded from macro 'EMIT'
QString toQString(const name &h) { \
        ^
/home/yegor/devel/snowman/src/nc/input/le/LeParser.cpp:161:1: warning: unused function 'toQString' [-Wunused-function]
/home/yegor/devel/snowman/src/nc/input/le/LeParser.cpp:126:5: note: expanded from macro 'EMIT_ALL'
    EMIT(FIXUP_HEADER, fixup_header) \
    ^
/home/yegor/devel/snowman/src/nc/input/le/LeParser.cpp:155:9: note: expanded from macro 'EMIT'
QString toQString(const name &h) { \
        ^
2 warnings generated.
```
Any ideas on how to fix this in a clean fashion?

File: src/nc/input/le/LeParser.cpp [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+        }
+        char le_sig[2];
+        if (!read(in, le_sig))
+            goto rewind;
+        if (memcmp(le_sig, "LE", 2))
+            goto rewind;
+        header_pos hpos = { .mz = mz, .le = mz + le_off };
+        return hpos;
+    }
+    return { -1, -1 };
```
Comment (yegord) [inline src/nc/input/le/LeParser.cpp]: This is not supported by MSVC 2010.

**Comment by yegord** [comment]
Okay, I've fixed the warnings.
Could you have a look at my changes?
If they look okay for you, I'll merge them.

**Comment by user7** [comment]
Ah, i see, packed is not even needed here, all structs are correct size.

All changes are good.

**Comment by yegord** [comment]
Merged. Thanks!

**Pull request closed without merging**